File: src/storage/repository.py
"""Repository for research sessions."""
from __future__ import annotations

from abc import ABC
from abc import abstractmethod
from typing import Final

from src.domain.session import ResearchSession
from src.storage.file import JsonFile


class Loadable(ABC):
    """Object that can load data."""

    @abstractmethod
    def load(self) -> tuple[ResearchSession, ...]:
        """Return all sessions."""
        ...


class Savable(ABC):
    """Object that can save data."""

    @abstractmethod
    def save(self, sessions: tuple[ResearchSession, ...]) -> None:
        """Persist all sessions."""
        ...


class SessionsRepository(Loadable, Savable):
    """Repository for research sessions."""

    def __init__(self, file: JsonFile) -> None:
        """Initialize with JSON file."""
        self._file: Final[JsonFile] = file

    def load(self) -> tuple[ResearchSession, ...]:
        """Return all sessions."""
        if not self._file.exists():
            return tuple()
        data = self._file.read()
        return tuple(
            ResearchSession.deserialize(s) for s in data.get("sessions", [])
        )

    def save(self, sessions: tuple[ResearchSession, ...]) -> None:
        """Persist all sessions."""
        self._file.write(
            {
                "version": "1.0.0",
                "sessions": [s.serialize() for s in sessions],
            }
        )

    def append(self, session: ResearchSession) -> None:
        """Add session to repository."""
        existing = self.load()
        self.save(existing + (session,))

    def find(self, identifier: str) -> ResearchSession | None:
        """Find session by identifier."""
        for session in self.load():
            if session.id() == identifier:
                return session
        return None

    def update(self, session: ResearchSession) -> None:
        """Update existing session."""
        sessions = self.load()
        updated = tuple(
            session if s.id() == session.id() else s for s in sessions
        )
        self.save(updated)

File: src/research/storage/repository.clj
(ns research.storage.repository
  (:refer-clojure :exclude [find load update])
  (:require [research.domain.session :as session]
            [research.storage.repository.read :as read]
            [research.storage.repository.write :as write])
  (:import (java.util Optional)))

(defprotocol Loadable
  "Object that can load sessions."
  (load [item] "Return session list."))

(defprotocol Savable
  "Object that can save sessions."
  (save [item list] "Persist session list."))

(defprotocol Mutable
  "Object that can update sessions."
  (append [item value] "Append session.")
  (find [item value] "Find session by id.")
  (update [item value] "Update session by id."))

(defrecord Repository [root]
  Loadable
  (load [_]
    (read/items root))
  Savable
  (save [_ items]
    (write/store root items))
  Mutable
  (append [item value]
    (let [items (load item)
          store (conj items value)]
      (save item store)))
  (find [item value]
    (let [items (load item)
          pick (first (filter #(= (session/id %) value) items))]
      (if pick (Optional/of pick) (Optional/empty))))
  (update [item value]
    (let [items (load item)
          store (mapv (fn [item]
                        (if (= (session/id item) (session/id value))
                          value
                          item))
                      items)]
      (save item store))))

(defn repo
  "Create repository from output path."
  [item]
  (->Repository item))
